### PR TITLE
net: udp_channel: integrate AF_UNIX support to the POSIX stack

### PR DESCRIPTION
**Pull request by havaker**
`posix_udp_channel`'s constructor assumed that its local address parameter will always be from `AF_INET`/ `AF_INET6` (`AF_UNSPECIFIED` gets converted to `AF_INET`). To support creating `AF_UNIX` udp_channels, additional checks were added before making inet specific calls in channel's constructor.

To get the information about local address, `posix_udp_channel` uses ancillary data residing in `struct msghdr`'s `msg_control`, `msg_controllen` fields. Their use was limited to inet only. Additionally, destination address returned by `udp_channel::receive` does not depend completely on the ancillary data, a fallback is provided for families other than inet.

`bind()` invoked in `posix_udp_channel`'s constructor body now receives the true address length - this is needed as unix domain addresses have no fixed width (in contrast to inet), and are not necessarily null-terminated.

Ref: https://github.com/scylladb/seastar/issues/1798

### CODE REVIEW

**Comment by nyh** [comment]
Before I go reviewing the details, I want to point out that the word "UDP", the "User Datagram Protocol", is part of the TCP/IP protocol suite, and should not be used to name other things such as unix-domain datagram sockets. So we should probably use the word "datagram", not "udp", in code which aims to support both IP and Unix-domain versions of the API.
Similarly we wouldn't use the word "TCP socket" to describe stream sockets in general, which may use TCP, Unix-domain, TLS, and so on.

File: src/net/posix-stack.cc [hunk 1]
```diff
@@ -769,21 +775,29 @@ class posix_udp_channel : public udp_channel_impl {
...
+            _closed(false) {
         auto sa = bind_address.is_unspecified() ? socket_address(inet_address(inet_address::family::INET)) : bind_address;
         file_desc fd = file_desc::socket(sa.u.sa.sa_family, SOCK_DGRAM | SOCK_NONBLOCK | SOCK_CLOEXEC, 0);
-        fd.setsockopt(SOL_IP, IP_PKTINFO, true);
-        if (engine().posix_reuseport_available()) {
-            fd.setsockopt(SOL_SOCKET, SO_REUSEPORT, 1);
+
+        if (eligible_to_ip_pktinfo(sa)) {
+            fd.setsockopt(SOL_IP, IP_PKTINFO, true);
+            if (engine().posix_reuseport_available()) {
```
Comment (nyh) [inline src/net/posix-stack.cc:801]: Why did you move this case inside the if(eligible)? is it because SO_REUSEPORT happens to work on the same type of sockets that support IP_PKTINFO? But the SOL_SOCKET instead of SOL_IP suggests it should have worked on Unix domain sockets as well (although perhaps not be especially useful?)

File: src/net/posix-stack.cc [hunk 2]
```diff
@@ -871,7 +871,14 @@ class posix_datagram : public udp_datagram_impl {
     posix_datagram(const socket_address& src, const socket_address& dst, packet p) : _src(src), _dst(dst), _p(std::move(p)) {}
     virtual socket_address get_src() override { return _src; }
     virtual socket_address get_dst() override { return _dst; }
-    virtual uint16_t get_dst_port() override { return _dst.port(); }
+    virtual uint16_t get_dst_port() override {
+        if (_dst.family() != AF_INET && _dst.family() != AF_INET6) {
+            std::stringstream ss;
+            ss << _dst;
+            throw std::runtime_error(format("get_dst_port() called on non-IP address: %s", ss.str()));
```
Comment (nyh) [inline src/net/posix-stack.cc:878]: I don't think you need this stringstream stuff, format should be able to do it itself. Something like:
```c++
throw std::runtime_error(format("get_dst_port() called on non-IP address: {}",_dst));
```

**Review by nyh — CHANGES_REQUESTED** [review]
Looks good, I just had a few minor comments.
I have a bigger problem with the usage of the name "udp" throughout the API, but I guess since the API already existed, we can let it remain with the original names and consider renaming later :-( But maybe if we don't rename things, we should at least add comments saying that udp_datagram, udp_channel, and other types, work not just for UDP (IP) but also for unix-domain datagram sockets.

**Comment by havaker** [comment]
> Before I go reviewing the details, I want to point out that the word "UDP", the "User Datagram Protocol", is part of the TCP/IP protocol suite, and should not be used to name other things such as unix-domain datagram sockets. So we should probably use the word "datagram", not "udp", in code which aims to support both IP and Unix-domain versions of the API. Similarly we wouldn't use the word "TCP socket" to describe stream sockets in general, which may use TCP, Unix-domain, TLS, and so on.

Thanks for the clear explanation! The line between *datagram* and *udp* was quite blurred in my mind.

I guess that it does not make sense to merge non-udp functionality to the `net::udp_channel`. A separate API  (e.g. `unix_dgram_channel`) or a generalization of the existing one (e.g. `dgram_channel`) would be a better fit. I'll close this pull request, and describe the problem in an appropriate issue.

File: src/net/posix-stack.cc (hunk 1 above)
Comment (havaker) [reply src/net/posix-stack.cc:801]: > is it because SO_REUSEPORT happens to work on the same type of sockets that support IP_PKTINFO

yes, i should have named `eligible_to_ip_pktinfo` somehow better

File: src/net/posix-stack.cc [hunk 3]
```diff
@@ -769,21 +775,29 @@ class posix_udp_channel : public udp_channel_impl {
...
+    }
     pollable_fd _fd;
     socket_address _address;
     recv_ctx _recv;
     send_ctx _send;
     bool _closed;
 public:
     posix_udp_channel(const socket_address& bind_address)
-            : _closed(false) {
+            : _recv(eligible_to_ip_pktinfo(bind_address) || bind_address.is_unspecified()),
```
[the remark this answers is not in the document]
Comment (havaker) [reply src/net/posix-stack.cc:794]: It doesn't, they don't have the same argument

[the remark this answers is not in the document]
Comment (nyh) [reply src/net/posix-stack.cc:794]: What do you mean don't have the same argument? Don't both check bind_address?

**Comment by nyh** [comment]
> > Before I go reviewing the details, I want to point out that the word "UDP", the "User Datagram Protocol", is part of the TCP/IP protocol suite, and should not be used to name other things such as unix-domain datagram sockets. So we should probably use the word "datagram", not "udp", in code which aims to support both IP and Unix-domain versions of the API. Similarly we wouldn't use the word "TCP socket" to describe stream sockets in general, which may use TCP, Unix-domain, TLS, and so on.
> 
> Thanks for the clear explanation! The line between _datagram_ and _udp_ was quite blurred in my mind.
> 
> I guess that it does not make sense to merge non-udp functionality to the `net::udp_channel`. A separate API (e.g. `unix_dgram_channel`) or a generalization of the existing one (e.g. `dgram_channel`) would be a better fit. I'll close this pull request, and describe the problem in an appropriate issue.

As I noted in the end, given that your changes are relatively small, and the API with the "UDP" in its name already exists (udp_channel, udp_datagram, etc.), maybe it does make some sense to tack on the extra functionality on the existing API and only later consider renaming it. If we rename everything today it will be a backward-compatibility mess. I am not familiar enough with the API to understand if we can use some sort of inhertence to support both the more general "datagram" and more specific "udp" and "unix domain datagram socket" on it. I'll let you think about it :-) I just wanted to bring this issue up.

File: src/net/posix-stack.cc [hunk 4]
```diff
@@ -730,15 +730,21 @@ class posix_udp_channel : public udp_channel_impl {
         char* _buffer;
         cmsg_with_pktinfo _cmsg;
 
-        recv_ctx() {
+        recv_ctx(bool use_pktinfo) {
```
[the remark this answers is not in the document]
Comment (havaker) [reply src/net/posix-stack.cc:733]: > why in a unix-domain datagram socket that has non-empty ancillary data you decide to zero msg_controllen and basically hide information that exists.

Hmm can a unix-domain datagram socket have a non-empty ancillary data if I never request them?
I decided to zero `msg_controllen` before the call to `recvmsg`, and thus I'm asking the kernel not to copy not needed ancillary data to the userspace.

From [man 2 recvmsg](https://www.man7.org/linux/man-pages/man2/recvmsg.2.html):
>       When recvmsg() is called,
>       msg_controllen should contain the length of the available buffer
>       in msg_control; upon return from a successful call it will
>       contain the length of the control message sequence.

File: src/net/posix-stack.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (havaker) [reply src/net/posix-stack.cc:794]: One is chcecking `bind_address`, the second one `sa`. Which is defined below as:
```c++
auto sa = bind_address.is_unspecified() ? socket_address(inet_address(inet_address::family::INET)) : bind_address;
```

[the remark this answers is not in the document]
Comment (havaker) [reply src/net/posix-stack.cc:794]: `is_unspecified()` boils down to `addr.family == AF_UNSPECIFIED`. This address family isn't eligible to receive `IP_PKTINFO` sockopt, so I didn't want to confuse the reader by containing `is_unspecified()` in `eligible_to_ip_pktinfo` method.

I feel like this boils down to a naming issue, maybe `is_ip_addr` would be a better choice.

**Comment by avikivity** [comment]
We could rename udp_channel to datagram_channel and deprecate the old name.
